src/main/java/org/stellar/sdk/Sep10Challenge.java

```diff
@@ -40,4 +48,344 @@ public static String newChallenge(
+    // decode the received input as a base64-urlencoded XDR representation of Stellar transaction envelope
+    Transaction transaction = Transaction.fromEnvelopeXdr(challengeXdr, network);
+
+    // verify that transaction source account is equal to the server's signing key
+    if (!serverAccountId.equals(transaction.getSourceAccount())) {
+      throw new InvalidSep10ChallengeException("Transaction source account is not equal to server's account.");
+    }
+
+    // verify that transaction sequenceNumber is equal to zero
+    if (transaction.getSequenceNumber() != 0L) {
+      throw new InvalidSep10ChallengeException("The transaction sequence number should be zero.");
+    }
+
+    // verify that transaction has time bounds set, and that current time is between the minimum and maximum bounds.
+    if (transaction.getTimeBounds() == null) {
+      throw new InvalidSep10ChallengeException("Transaction requires timebounds.");
+    }
+
+    long maxTime = transaction.getTimeBounds().getMaxTime();
+    long minTime = transaction.getTimeBounds().getMinTime();
+    if (maxTime == 0L) {
+      throw new InvalidSep10ChallengeException("Transaction requires non-infinite timebounds.");
+    }
+
+    long currentTime = System.currentTimeMillis() / 1000L;
+    if (currentTime < minTime || currentTime > maxTime) {
+      throw new InvalidSep10ChallengeException("Transaction is not within range of the specified timebounds.");
+    }
+
+    // verify that transaction contains a single Manage Data operation and its source account is not null
+    if (transaction.getOperations().length != 1) {
+      throw new InvalidSep10ChallengeException("Transaction requires a single ManageData operation.");
+    }
+    Operation operation = transaction.getOperations()[0];
+    if (!(operation instanceof ManageDataOperation)) {
+      throw new InvalidSep10ChallengeException("Operation type should be ManageData.");
+    }
+    ManageDataOperation manageDataOperation = (ManageDataOperation) operation;
+
+    // verify that transaction envelope has a correct signature by server's signing key
+    String clientAccountId = manageDataOperation.getSourceAccount();
+    if (clientAccountId == null) {
+      throw new InvalidSep10ChallengeException("Operation should have a source account.");
+    }
+
+    // verify manage data value
+    if (manageDataOperation.getValue().length != 64) {
+      throw new InvalidSep10ChallengeException("Random nonce encoded as base64 should be 64 bytes long.");
+    }
+
+    BaseEncoding base64Encoding = BaseEncoding.base64();
+    byte[] nonce;
+    try {
+      nonce = base64Encoding.decode(new String(manageDataOperation.getValue()));
+    } catch (IllegalArgumentException e) {
+      throw new InvalidSep10ChallengeException("Failed to decode random nonce provided in ManageData operation.", e);
```

Comment:
@leighmcculloch I added a bit of code here, so it looks better?

Reply:
Looks great.